analyzers/src/SonarAnalyzer.VisualBasic/Rules/UriShouldNotBeHardcoded.cs

```diff
@@ -21,38 +21,29 @@
-        : UriShouldNotBeHardcodedBase<ExpressionSyntax, LiteralExpressionSyntax,
-            SyntaxKind, BinaryExpressionSyntax, ArgumentSyntax, VariableDeclaratorSyntax>
+    public sealed class UriShouldNotBeHardcoded : UriShouldNotBeHardcodedBase<SyntaxKind, LiteralExpressionSyntax, ArgumentSyntax>
     {
-        private static readonly DiagnosticDescriptor Rule =
-            DescriptorFactory.Create(DiagnosticId, MessageFormat);
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
-        protected override SyntaxKind StringLiteralSyntaxKind => SyntaxKind.StringLiteralExpression;
-
-        protected override SyntaxKind[] StringConcatenateExpressions =>
-            new[]
-            {
-                SyntaxKind.AddExpression,
-                SyntaxKind.ConcatenateExpression
-            };
-
+        protected override ILanguageFacade<SyntaxKind> Language => VisualBasicFacade.Instance;
         protected override GeneratedCodeRecognizer GeneratedCodeRecognizer => VisualBasicGeneratedCodeRecognizer.Instance;
+        protected override SyntaxKind[] StringConcatenateExpressions => new[] { SyntaxKind.AddExpression, SyntaxKind.ConcatenateExpression };
+        protected override SyntaxKind[] InvocationOrObjectCreationKind => new[] { SyntaxKind.InvocationExpression, SyntaxKind.ObjectCreationExpression };
 
-        protected override string GetLiteralText(LiteralExpressionSyntax literalExpression) => literalExpression?.Token.ValueText;
-
-        protected override ExpressionSyntax GetLeftNode(BinaryExpressionSyntax binaryExpression) => binaryExpression.Left;
-
-        protected override ExpressionSyntax GetRightNode(BinaryExpressionSyntax binaryExpression) => binaryExpression.Right;
+        protected override string GetLiteralText(LiteralExpressionSyntax literalExpression) =>
+            literalExpression?.Token.ValueText;
 
-        protected override bool IsInvocationOrObjectCreation(SyntaxNode node) =>
-            node.IsKind(SyntaxKind.InvocationExpression)
-            || node.IsKind(SyntaxKind.ObjectCreationExpression);
-
-        protected override int? GetArgumentIndex(ArgumentSyntax argument) =>
-            (argument?.Parent as ArgumentListSyntax)?.Arguments.IndexOf(argument);
-
-        protected override string GetDeclaratorIdentifierName(VariableDeclaratorSyntax declarator) =>
-            declarator.Names.FirstOrDefault()?.Identifier.ValueText;
+        protected override SyntaxNode GetRelevantAncestor(SyntaxNode node)
+        {
+            var parameter = node.FirstAncestorOrSelf<ParameterSyntax>();
+            if (parameter != null)
+            {
+                return parameter;
+            }
+            var variableDeclarator = node.FirstAncestorOrSelf<VariableDeclaratorSyntax>();
+            if (variableDeclarator != null)
+            {
+                return variableDeclarator;
+            }
+            return null;
```

Comment:
Yes, it's correct, I was not sure if there were other ancestors that we could fit into this logic, and then I kept it as is. I think I prefer the slimmer version you suggested. Eventually, it might get expanded in the future.